File: Data-Structure-I/1-Array/P118.cpp
// Date: Sat, Oct 2, 2021
// Redo (Did it in Dynamic Programming I)
// 118. Pascal's Triangle (https://leetcode.com/problems/pascals-triangle/)

// time complexity: O(n^2) <- T(n) = 1 + 2 + 3 + ... + n where n is numRows
// extra space complexity: O(1)
vector<vector<int>> generate(int numRows) {
	vector<vector<int>> ans;
	ans.push_back({1});
	for (int i = 1; i < numRows; i++) {
		vector<int> cur(i + 1);
		for (int j = 0; j < i + 1; j++) {
			if (j == 0 || j == i)
				cur[j] = 1;
			else 
				cur[j] = ans[i-1][j-1] + ans[i-1][j];
		}
		ans.push_back(cur);
	}
	return ans;
}

File: LeetCoding-Challenge/July-2022/P118.java
package leetcodingchallenge.july2022;

import java.util.ArrayList;
import java.util.List;

/**
 * Date: July 19, 2022 - Redo (already solve in C++ since Sept 19, 2021)
 * 118. Pascal's Triangle
 * https://leetcode.com/problems/pascals-triangle/
 */

public class P118 {
    // time: O(numRows^2)
    // space: O(1)
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> ans = new ArrayList<>();
        for (int row = 0; row < numRows; row++) {
            List<Integer> list = new ArrayList<>();
            if (row == 0) {
                list.add(1);
            } else {
                List<Integer> prev = ans.get(row - 1);
                for (int col = 0; col < row + 1; col++) {
                    int val = 0;
                    val += col > 0 ? prev.get(col - 1) : 0;
                    val += col < row ? prev.get(col) : 0;
                    list.add(val);
                }
            }
            ans.add(list);
        }
        return ans;
    }
}
